### PR TITLE
[DROOLS-7635] ansible-rulebook : Raise an error when a condition compares incompatible types

### PR DESCRIPTION
**Pull request by tkobayas**
This draft PR requires
https://github.com/apache/incubator-kie-drools/pull/6104
to be built beforhand.

Cross repo PR check doesn't work for now.

### CODE REVIEW

**Comment by tkobayas** [comment]
@mariofusco with https://github.com/apache/incubator-kie-drools/pull/6104 , now alpha index and node sharing is supported.

Limitation: in case of  node sharing, the error message has only one ruleSetName/ruleName/conditionExpression. It means that when a user fixes a condition based on the error message, they may hit the similar issue in another condition next time. It might be invasive to "combine" error messages upon node sharing in drools code base. We may ask Madhu if it's important.

**Comment by tkobayas** [comment]
The point I mentioned at https://github.com/kiegroup/drools-ansible-rulebook-integration/pull/112#discussion_r1778056780 still remains
- Alpha index skips unmatched AlphaNodes and jumps to the right AlphaNode. That means, **when an event doesn't match any AlphaNode because of type mismatch** , the custom predicate would not be executed. This is actually the case that users want to read the error message ... "No rule fires, I don't know if constraints are wrong". I think we would need to add the check logic in `CompositeObjectSinkAdapter.propagateAssertObject`.

I'll need to consider further.

File: drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/domain/constraints/RulebookOperator.java [hunk 1]
```diff
@@ -19,17 +21,42 @@ default ConstraintOperator asConstraintOperator() {
...
+
+    default void setConditionContext(RuleGenerationContext ruleContext, Map<?, ?> expression) {
+        if (this instanceof OperatorWrapper operatorWrapper) {
+            operatorWrapper.setConditionContext(ruleContext, expression);
+        } else {
+            // do nothing
+        }
+    }
 
     class OperatorWrapper implements RulebookOperator {
```
Comment (mariofusco) [inline drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/domain/constraints/RulebookOperator.java:56]: This wrapper seems no longer necessary, couldn't `RulebookConstraintOperator` directly implement `RulebookOperator`?

**Comment by tkobayas** [comment]
**DO-NOT-MERGE until https://github.com/apache/incubator-kie-drools/pull/6104 will be merged and the commit is synced to `kiegroup/drools`**

**Review by mariofusco — APPROVED** [review]

File: drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/domain/constraints/RulebookOperator.java (hunk 1 above)
Comment (tkobayas) [reply drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/domain/constraints/RulebookOperator.java]: Thanks, done.

**Comment by rgdoliveira** [comment]
> **DO-NOT-MERGE until [apache/incubator-kie-drools#6104](https://github.com/apache/incubator-kie-drools/pull/6104) will be merged and the commit is synced to `kiegroup/drools`**

@tkobayas FYI I'm doing the sync in PR https://github.com/kiegroup/drools/pull/80

**Pull request merged**